// ng-forward/es6/util/helpers.js
const SNAKE_CASE_REGEXP = /[A-Z]/g;
export function ucFirst(word) {
    return `${word.charAt(0).toUpperCase()}${word.substring(1)}`;
}
export function dashToCamel(dash) {
    let words = dash.split('-');
    return `${words.shift()}${words.map(ucFirst).join('')}`;
}
export function dasherize(name, separator = '-') {
    return name.replace(SNAKE_CASE_REGEXP, (letter, pos) => {
        return `${(pos ? separator : '')}${letter.toLowerCase()}`;
    });
}
export function snakeCase(name, separator = '-') {
    return name.replace(SNAKE_CASE_REGEXP, (letter, pos) => {
        return `${(pos ? separator : '')}${letter.toLowerCase()}`;
    });
}
export function flatten(items) {
    let resolved = [];
    for (let item of items) {
        if (Array.isArray(item)) {
            resolved.push(...flatten(item));
        }
        else {
            resolved.push(item);
        }
    }
    return resolved;
}
export function createConfigErrorMessage(target, ngModule, message) {
    return `Processing "${target.name}" in "${ngModule.name}": ${message}`;
}
