// ng-forward/es6/events/events.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive } from '../decorators/directive';
import { Inject } from '../decorators/inject';
import parseSelector from '../util/parse-selector';
import { dasherize } from '../util/helpers';
let events = new Set([
    'click',
    'dblclick',
    'mousedown',
    'mouseup',
    'mouseover',
    'mouseout',
    'mousemove',
    'mouseenter',
    'mouseleave',
    'keydown',
    'keyup',
    'keypress',
    'submit',
    'focus',
    'blur',
    'copy',
    'cut',
    'paste',
    'change',
    'dragstart',
    'drag',
    'dragenter',
    'dragleave',
    'dragover',
    'drop',
    'dragend',
    'error',
    'input',
    'load',
    'wheel',
    'scroll'
]);
function resolve() {
    let directives = [];
    events.forEach(event => {
        const selector = `[(${dasherize(event)})]`;
        let EventHandler = class {
            constructor($parse, $element, $attrs, $scope) {
                this.$element = $element;
                this.$scope = $scope;
                let { name: attrName } = parseSelector(selector);
                this.expression = $parse($attrs[attrName]);
                $element.on(event, e => this.eventHandler(e));
                $scope.$on('$destroy', () => this.onDestroy());
            }
            eventHandler($event = {}) {
                let detail = $event.detail;
                if (!detail && $event.originalEvent && $event.originalEvent.detail) {
                    detail = $event.originalEvent.detail;
                }
                else if (!detail) {
                    detail = {};
                }
                this.expression(this.$scope, Object.assign(detail, { $event }));
                this.$scope.$applyAsync();
            }
            onDestroy() {
                this.$element.off(event);
            }
        };
        EventHandler = __decorate([
            Directive({ selector }),
            Inject('$parse', '$element', '$attrs', '$scope'), 
            __metadata('design:paramtypes', [Function, Object, Object, Object])
        ], EventHandler);
        directives.push(EventHandler);
    });
    return directives;
}
function add(...customEvents) {
    customEvents.forEach(event => events.add(event));
}
export default { resolve, add };
