// ng-forward/es6/classes/metastore.js
export default class Metastore {
    constructor(namespace) {
        this.namespace = namespace;
    }
    _map(obj, key) {
        if (!Reflect.hasOwnMetadata(this.namespace, obj, key)) {
            Reflect.defineMetadata(this.namespace, new Map(), obj, key);
        }
        return Reflect.getOwnMetadata(this.namespace, obj, key);
    }
    get(key, obj, prop) {
        return this._map(obj, prop).get(key);
    }
    set(key, value, obj, prop) {
        this._map(obj, prop).set(key, value);
    }
    has(key, obj, prop) {
        return this._map(obj, prop).has(key);
    }
    push(key, value, obj, prop) {
        if (!this.has(key, obj, prop)) {
            this.set(key, [], obj, prop);
        }
        let store = this.get(key, obj, prop);
        if (!Array.isArray(store)) {
            throw new Error('Metastores can only push metadata to array values');
        }
        store.push(value);
    }
    merge(key, value, obj, prop) {
        let previous = this.get(key, obj, prop) || {};
        let mergedObj = Object.assign({}, previous, value);
        this.set(key, mergedObj, obj, prop);
    }
    forEach(callbackFn, obj, prop) {
        this._map(obj, prop).forEach(callbackFn);
    }
}
