// ng-forward/cjs/util/jqlite-extensions.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

var _getInjectableName = require('./get-injectable-name');

(function extendJQLite(proto) {
    Object.defineProperties(proto, {
        nativeElement: {
            get: function get() {
                return this[0];
            }
        },
        componentInstance: {
            get: function get() {
                if (this._componentInstance) return this._componentInstance;
                var isolateScope = this.isolateScope();
                this._componentInstance = isolateScope && isolateScope['ctrl'] || null;
                return this._componentInstance;
            }
        },
        componentViewChildren: {
            get: function get() {
                return [].concat(_toConsumableArray(this.children())).map(function (child) {
                    return angular.element(child);
                });
            }
        },
        getLocal: {
            value: function value(injectable) {
                return (this.injector() || this.inheritedData('$injector')).get((0, _getInjectableName.getInjectableName)(injectable));
            }
        },
        query: {
            value: function value(predicate, scope) {
                throw new Error('Not yet implemented in ng-forward.');
            }
        },
        queryAll: {
            value: function value(predicate, scope) {
                throw new Error('Not yet implemented in ng-forward.');
            }
        },
        getDirectiveInstance: {
            value: function value(index) {
                throw new Error('Not yet implemented in ng-forward.');
            }
        },
        triggerEventHandler: {
            value: function value(eventName, eventObj) {
                throw new Error('Not yet implemented in ng-forward.');
            }
        },
        inject: {
            value: function value(type) {
                throw new Error('Not yet implemented in ng-forward.');
            }
        },
        hasDirective: {
            value: function value(type) {
                throw new Error('Not yet implemented in ng-forward.');
            }
        }
    });
})(angular.element.prototype);
exports['default'] = angular.element;
module.exports = exports['default'];
